arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/raise/RaiseAccumulate.kt

```diff
@@ -7,574 +7,443 @@ import arrow.core.mapOrAccumulate
+  val d = recover({ action4(RaiseAccumulate(this)) }) { error.addAll(it); EmptyValue }
+  val e = recover({ action5(RaiseAccumulate(this)) }) { error.addAll(it); EmptyValue }
+  val f = recover({ action6(RaiseAccumulate(this)) }) { error.addAll(it); EmptyValue }
+  val g = recover({ action7(RaiseAccumulate(this)) }) { error.addAll(it); EmptyValue }
+  val h = recover({ action8(RaiseAccumulate(this)) }) { error.addAll(it); EmptyValue }
+  val i = recover({ action9(RaiseAccumulate(this)) }) { error.addAll(it); EmptyValue }
+  error.toNonEmptyListOrNull()?.let { raise(it) }
+  return block(unbox(a), unbox(b), unbox(c), unbox(d), unbox(e), unbox(f), unbox(g), unbox(h), unbox(i))
 }
 
 /**
- * Accumulate the errors obtained by executing the [block]
- * over every element of [list] using the given [semigroup].
+ * Transform every element of [list] using the given [transform], or accumulate all the occurred errors using [combine].
  */
 @RaiseDSL
 public inline fun <R, A, B> Raise<R>.mapOrAccumulate(
-  semigroup: Semigroup<@UnsafeVariance R>,
   list: Iterable<A>,
-  @BuilderInference block: Raise<R>.(A) -> B
+  combine: (R, R) -> R,
+  @BuilderInference transform: RaiseAccumulate<R>.(A) -> B
 ): List<B> {
-  // this could be implemented using [zipOrAccumulate],
-  // but we can have a faster implementation using [buildList]
   var error: Any? = EmptyValue
-  val results = buildList {
-    list.forEach {
-      fold<R, B, Unit>({
-        block(it)
-      }, { newError ->
-        error = semigroup.emptyCombine(error, newError)
-      }, {
-        add(it)
-      })
-    }
-  }
-  return when (val e = EmptyValue.unbox<R>(error)) {
-    null -> results
-    else -> raise(e)
+  val results = ArrayList<B>(list.collectionSizeOrDefault(10))
+  for (item in list) {
+    fold<NonEmptyList<R>, B, Unit>(
+      { transform(RaiseAccumulate(this), item) },
+      { errors -> error = combine(error, errors.reduce(combine), combine) },
+      { results.add(it) }
+    )
   }
+  return if (error === EmptyValue) results else raise(unbox<R>(error))
 }
 
 /**
- * Accumulate the errors obtained by executing the [block]
- * over every element of [list].
+ * Accumulate the errors obtained by executing the [transform] over every element of [list].
  */
 @RaiseDSL
 public inline fun <R, A, B> Raise<NonEmptyList<R>>.mapOrAccumulate(
   list: Iterable<A>,
-  @BuilderInference crossinline block: Raise<R>.(A) -> B
-): List<B> =
-  mapOrAccumulate(Semigroup.nonEmptyList(), list) { elt -> mapErrorNel { block(elt) } }
+  @BuilderInference transform: RaiseAccumulate<R>.(A) -> B
+): List<B> {
+  val error = mutableListOf<R>()
+  val results = ArrayList<B>(list.collectionSizeOrDefault(10))
+  for (item in list) {
+    fold<NonEmptyList<R>, B, Unit>(
+      { transform(RaiseAccumulate(this), item) },
+      { errors -> error.addAll(errors) },
+      { results.add(it) }
+    )
+  }
+  return error.toNonEmptyListOrNull()?.let { raise(it) } ?: results
```

Comment:
The fact that this can be defined now once and for all in a nice way fills me with joy :)